### DISCUSSION TITLE
How to refetch specific fragment based on user interactions?

### DISCUSSION THREAD

**Discussion by 2coo** [dotansimha/graphql-code-generator#9657] [Q&A]
I just saw [the blog](https://the-guild.dev/blog/unleash-the-power-of-fragments-with-graphql-codegen) and I am wondering how to refetch (reload) `FriendList_QueryFragment` when the user clicks refresh button? instead of rerunning the entire big single query operation?

**Comment by matthewhausman** [comment]
Is there any way to do this?

**Reply by saihaj** [reply]
I know in Relay they have [refetchable fragments](https://relay.dev/docs/api-reference/use-refetchable-fragment/), don't think any other GraphQL client supports that.

**Reply by matthewhausman** [reply]
so it seems that relay constructs an additional query... I did a similar approach. Populate the query with initial data and now you have control of refetching

```ts
const document = graphql(/* GraphQL */ `
  query industriesHome {
    ...IndustriesQueryFragmentHome
  }
`)

export function Industries(props: { query: FragmentType<typeof IndustriesQueryFragmentHome> }) {
   const q = useQuery(
    useMemo(
      () => ({
        ...queryOptionGenerator({ document, args: {} }),
        initialData: props.query,
        gcTime: Infinity,
        staleTime: Infinity,
      }),
      [props.query]
    )
  )

  if (!q.data) return null

  const data = getFragmentData(IndustriesQueryFragmentHome, q.data)
```

**Reply by matthewhausman** [reply]
quite a beautiful pattern, having a great time with this and fragment parameters

**Reply by 2coo** [reply]
@matthewhausman

Does it refetch the whole Query, instead of refetching just the fragment?

**Reply by matthewhausman** [reply]
see how im passing the query data that I got originally to the initialData of my fragment's query?

**Reply by matthewhausman** [reply]
In the docs they use the same approach. pass the query to child components.

**Comment by 2coo** [comment]
[HoudiniGraphQL](https://houdinigraphql.com/) is my best client with relay specs so far.
